password_enhancements.module

```diff
@@ -0,0 +1,471 @@
+  $password = $form_state->getValue('pass');
+  if (!empty($password)) {
+    $roles = $form_state->getValue('roles') ?? [];
+
+    // Try to load policy by it's role and priority.
+    try {
+      $entity_type_manager = Drupal::entityTypeManager();
+      /** @var \Drupal\password_enhancements\Entity\Storage\PolicyEntityStorageInterface $password_policy_storage */
+      $password_policy_storage = $entity_type_manager->getStorage('password_enhancements_policy');
+      $policy = $password_policy_storage->loadByRoleAndPriority($roles);
+
+      /** @var \Drupal\password_enhancements\Entity\Storage\ConstraintEntityStorageInterface $constraint_storage */
+      $constraint_storage = $entity_type_manager->getStorage('password_enhancements_constraint');
+      $constraints = $constraint_storage->loadByRoles($roles);
+    }
+    catch (Exception $e) {
+      Drupal::service('logger.password_enhancements')->exception($e);
+      $form_state->setError($form['account']['pass'], t('Not possible to validate the password: failed to load policy.'));
+    }
+
+    // If no policies and constraints are defined then we have nothing to do
+    // here.
+    if (empty($policy) && empty($constraints)) {
+      return;
+    }
+
+    // Get constraints and count the minimum required constraints.
+    $constraints_count = count($constraints);
+    $minimum_required_constraints = $policy->getMinimumRequiredConstraints();
+    $minimum_required_constraints = $constraints_count < $minimum_required_constraints ? $constraints_count : $minimum_required_constraints;
+
+    // Validate password against the defined constraints.
+    $error_message_collection = [
+      'optional' => [],
+      'required' => [],
+    ];
+    $valid_required = 0;
+    $valid_optional = 0;
+    $plugin_manager = Drupal::service('plugin.manager.password_constraint');
+    foreach ($constraints as $constraint) {
+      /** @var \Drupal\password_enhancements\Plugin\PasswordConstraintPluginInterface $plugin_instance */
+      $plugin_instance = $plugin_manager->createInstance($constraint->getType(), $constraint->getConfiguration());
+      $is_required = $constraint->isRequired();
+      $is_valid = $plugin_instance->validate($password);
+
+      if ($is_valid && $is_required) {
+        $valid_required++;
+      }
+      elseif ($is_valid) {
+        $valid_optional++;
+      }
+      else {
+        $error_message_collection[$is_required ? 'required' : 'optional'][] = $plugin_instance->getErrorMessage();
+      }
+    }
+
+    // Collect error messages.
+    $error_messages = [];
+    if (($valid_required + $valid_optional) < $minimum_required_constraints) {
+      $error_messages = array_merge($error_message_collection['required'], $error_message_collection['optional']);
+    }
+    elseif ($valid_required === 0 && $error_message_collection['required'] === [] && $valid_optional < $minimum_required_constraints) {
+      $error_messages = &$error_message_collection['optional'];
```

Comment:
Pass by reference is still dangerous. IMO the amount of memory you spare with this does not worth it.

Reply:
Not if you use it correctly, although both are defined locally, so it would not do any harm either way.